GFM Model "REGFM_A1"

lvanfretti opened the pull request:
Implementation of variants of the REGFM_A1 model of a GFM based on [UNIFI Report](https://www.wecc.org/sites/default/files/documents/products/2024/Model%20Specification%20of%20Droop-Controlled%20Grid-Forming%20Inverters-REGFM_A1.pdf) and related [paper](https://ieeexplore.ieee.org/document/10470420)

Issues to be solved:

- [ ] Small initialization inconsistencies (@lvanfretti in progress).
- [ ] Needs to implement fault current limiter (@lvanfretti in progress)
- [x] ~~Needs validation against PSS/E (help wanted).~~ (dropped)

lvanfretti committed:
REGFM_A1 Model Implementation

Initial commit of the REGFM_A1 grid forming converter model.
Needs debugging of initialization within control loops

lvanfretti committed:
Correct signal scaling

Fix signal scaling within packaged components

lvanfretti committed:
Corrected passing of initial voltage into the Q-v block

lvanfretti committed:
Amending issue with variability

Reverted unnecessary change
Improved test example

lvanfretti committed:
Small improvements and renaming, etc

lvanfretti commented:
@dietmarw there is an interesting error coming out of the syntax checker
```
syntax error on line 4 of OpenIPSL/Electrical/GFCs/GFMs/LimitersAndCntrlSchemes/P_f_droop_scheme.mo: expected identifier but received `break`
syntax error on line 4 of OpenIPSL/Electrical/GFCs/GFMs/LimitersAndCntrlSchemes/Q_v_droop_scheme.mo: expected identifier but received `break`
```
While this is legal in Dymola, I am wondering why MapleSim's syntax checker is marking it as an error?

dietmarw commented:
In general, there is quite a list of issues with this PR I can comment on then, but I thought I would wait until you have come so far that this no longer has the draft status.

dietmarw commented:
ONe suggestion to do before you change this from draft to real is to run a style check on the new subpackage and adjust the naming conventions accordingly.

dietmarw commented:
<details><summary>Style check:</summary>
Library style check log
OpenIPSL.Electrical.GFCs.GFMs
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.REGFM_A1_Light
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.REGFM_A1_incl_Limiters
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.BasicComponentsAndCntrlSchemes
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.BasicComponentsAndCntrlSchemes.low_pass_filter
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.BasicComponentsAndCntrlSchemes.pf_drp
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.BasicComponentsAndCntrlSchemes.pf_droop_scheme_nolims
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.BasicComponentsAndCntrlSchemes.qv_drp
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.BasicComponentsAndCntrlSchemes.qv_droop_scheme_nolims
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.BasicComponentsAndCntrlSchemes.VSIO_for_GFM
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.LimitersAndCntrlSchemes
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.LimitersAndCntrlSchemes.Plimiter
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.LimitersAndCntrlSchemes.P_f_droop_scheme
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.LimitersAndCntrlSchemes.Qlimiter
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.LimitersAndCntrlSchemes.Q_v_droop_scheme
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.BasicComponents
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.BasicComponents.LPFilt
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.BasicComponents.Pfbasic
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.BasicComponents.P_f_droop_scheme
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.BasicComponents.qv_droop_scheme_qvflag_1_vflag_1
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.BasicComponents.qv_droop_scheme_qvflag_1_vflag_0
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.BasicComponents.qv_droop_scheme_qvflag_0_vflag_1
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.BasicComponents.qv_droop_scheme_qvflag_0_vflag_0
Documentation missing
Bad class name Name of model shall start with capital letter. Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.Limiters
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.Limiters.Plimiter
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.Limiters.Qlimiter
Documentation missing
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.REGF_Basic
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.REGF_Basic.Components_unpackaged_system_test
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.REGF_Basic.Packaged_system_test
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.REGFM_A1_with_Limiters
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.REGFM_A1_with_Limiters.Components_unpackaged_system_test
Documentation missing
Bad class name Name shall not contain '_'.
OpenIPSL.Electrical.GFCs.GFMs.TestComponents.REGFM_A1_with_Limiters.Packaged_system_test
Documentation missing
Bad class name Name shall not contain '_'.
</details>

gkhalley commented:
Please let me know what you would like tested in PSSE. I suspect REGFM_A1_with_limiters -> Packaged_system_test. I also get the same droop scheme errors in WSM. The REGF_Basic->Packaged_system_test works, but after 40 seconds the voltage starts to rise and by 48 seconds it is 1.2 pu (maybe it's the reason limiters are needed).

lvanfretti commented:
@gkhalley let's try testing on WSM once I've done all the cleanup that @dietmarw has pointed out, and the three issues listed at the top of the PR. I am trying to find time after classes are over to clean this up.

lvanfretti committed:
Update Qlimiter test example

lvanfretti committed:
Style Check Corrections

lvanfretti committed:
Refactoring to use base classes instead of "break" for compatibility

Maplesim checker is not updated to latest modelica language

lvanfretti committed:
Fixes after refactoring

lvanfretti committed:
Documentation Style Check Fixes

lvanfretti committed:
Remove block mistakenly added that broke q-v loop

lvanfretti committed:
Additional fixes

lvanfretti committed:
Completed changes for style check compliance

lvanfretti commented:
@dietmarw I've now gone through all the Dymola style check and made fixes. However, it looks like additional changes are needed for the CI tests to pass.

Here is the output from the style check in Dymola:

    <HEAD><TITLE>HTML style check log for OpenIPSL.Electrical.GFCs.GFMs</TITLE></HEAD>
<h3><font color="#008000" size=5>Library style check log</font></h3>
<li>OpenIPSL.Electrical.GFCs.GFMs.VoltageSources, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.VoltageSources.VSIOforGFM, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.REGFMA1Light, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.REGFMA1wPQLimiters, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes.Base, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes.Base.PFdrpCntrlSchemeBase, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes.Base.QVdrpCntrlSchemeBase, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes.LPFilter, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes.PFdrp, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes.PFdrpScheme, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes.QVdrp, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.BasicComponentsAndCntrlSchemes.QVdrpCntrlScheme, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.LimitersAndCntrlSchemes, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.LimitersAndCntrlSchemes.Plimiter, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.LimitersAndCntrlSchemes.PfdrpControlSchemePLim</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.LimitersAndCntrlSchemes.Qlimiter, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.LimitersAndCntrlSchemes.QVdrpCntrlSchemeQLim</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.BasicComponents, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.BasicComponents.LPFilt</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.BasicComponents.Pfbasic</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.BasicComponents.PFdrpScheme</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.BasicComponents.QVqvflag1vflag1, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.BasicComponents.QVqvflag1vflag0, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.BasicComponents.QVqvflag0vflag1, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.BasicComponents.QVqvflag0vflag0, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.Limiters, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.Limiters.Plimiter</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.Limiters.Qlimiter</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.REGFMA1LightEncapsulate, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.REGFMA1LightEncapsulate.CombiningComponents, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.REGFMA1LightEncapsulate.ComponentEncapsulation</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.REGFMA1withPQLimEncapsulate, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.REGFMA1withPQLimEncapsulate.CombiningComponents, <font color="#008000" size=3>Check ok</font><ul type=disc></ul></li>
<li><b>OpenIPSL.Electrical.GFCs.GFMs.REGFMA1.TestComponents.REGFMA1withPQLimEncapsulate.ComponentEncapsulation</b><ul type=disc><li><font color="#FF9900" size=3>(Possibly) insufficient documentation</font></li></ul></li>

lvanfretti committed:
Fixes for CI checks

lvanfretti committed:
Removing leftovers

dietmarw reviewed (changes requested):
This PR breaks with the convention that all classes should start with a capital letter (functions being the allowed exception), i.e., should be using PascalCase. In addition, there is an unfortunate overuse of underscore. Try to stick to PascalCase and avoid that.

dietmarw commented:
@lvanfretti Is this stil Draft mode or ready for review?

lvanfretti committed:
Deleting left-overs

dietmarw reviewed (dismissed):
Looks good now.

lvanfretti committed:
Moving component test models into the Tests Sub-package

lvanfretti committed:
Virtual Resistor Implementation for Current Limiting Strategy

Added virtual resistor component and test model.
Renamed directories that only indicated "limiters" to specify which kind of limiter, as the current limiters are being added.

lvanfretti committed:
Adding current limit calculation block and test example

lvanfretti committed:
Voltage source with virtual resistance input

Voltage source with virtual resistance input to test functionality of virtual resistor block in a grid model.

lvanfretti committed:
Test model - VS with Virtual Re input

Test model incorporating the virtual resistance component with a voltage source in a power grid.

lvanfretti committed:
fix package.order files
